### PR TITLE
feat(pipeline/api): replace IOC.asn with AutonomousSystem FK. Closes #770

### LINKED ISSUE #770: AS API does not diplay the name of the AS
In our new AS API, we only show the number of the AS, but not its actual name. Although the ASN is the actual identifier of the AS, adding the name to the responses would be useful.

### PR DESCRIPTION
**Pull request by drona-gyawali**
# Description

Changes include:
- Created `AutonomousSystem` model with `asn` (primary key) and optional `name`.
- Migrated existing `IOC.asn` data to `IOC.autonomous_system` FK.
- Updated extraction utilities, scoring, and feeds to use `autonomous_system`.
- Updated admin display to show ASN and AS name neatly.
- Created `ASNRepository` for consistent get-or-create logic.
- Added comprehensive tests covering:
    - ASNRepository edge cases (create, update, logging)
    - iocs_from_hits enrichment with AS info
    - Migration from ASN field to AutonomousSystem FK
    - Feeds API and scoring utilities integration

### Related issues
closes: #770 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Chore (refactoring, dependency updates, CI/CD changes, code cleanup, docs-only changes).

# Checklist

Please complete this checklist carefully. It helps guide your contribution and lets maintainers verify that all requirements are met.

### Formalities

- [x] I have read and understood the rules about [how to Contribute](https://intelowlproject.github.io/docs/GreedyBear/Contribute/) to this project.
- [x] I chose an appropriate title for the pull request in the form: `<feature name>. Closes #999`
- [x] My branch is based on `develop`.
- [x] The pull request is for the branch `develop`.
- [x] I have reviewed and verified any LLM-generated code included in this PR.

### Docs and tests

- [x] I documented my code changes with docstrings and/or comments.
- [x] I have checked if my changes affect user-facing behavior that is described in the [docs](https://intelowlproject.github.io/docs/GreedyBear/Introduction/). If so, I also created a pull request in the [docs repository](https://github.com/intelowlproject/docs).
- [x] Linter (`Ruff`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowlproject.github.io/docs/GreedyBear/Contribute/#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [x] I have added tests for the feature/bug I solved.
- [x] All the tests gave 0 errors.

### GUI changes

Ignore this section if you did not make any changes to the GUI.

- [ ] I have provided a screenshot of the result in the PR.
- [ ] I have created new frontend tests for the new component or updated existing ones.
  
# Review process

- We encourage you to create a draft PR first, even when your changes are incomplete. This way you refine your code while we can track your progress and actively review and help.
- If you think your draft PR is ready to be reviewed by the maintainers, click the corresponding button. Your draft PR will become a real PR.
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem.
- Every time you make changes to the PR and you think the work is done, you should explicitly ask for a review. After receiving a "change request", address the feedback and click "request re-review" next to the reviewer's profile picture at the top right.

### CODE REVIEW

File: tests/test_migrations.py [hunk 1]
```diff
@@ -119,3 +119,77 @@ def test_log4pot_kept_if_has_iocs(self):
             hp_new.objects.filter(name="Log4pot").exists(),
             "Log4pot with IOCs should NOT be deleted",
         )
+
+    @tag("migration")
+    class TestIocAsnToAutonomousSystem(MigrationTestCase):
```
Comment (regulartim) [inline tests/test_migrations.py]: I think this class is nested inside `TestRemoveUnusedLog4pot`, therefore it wont run any tests.

File: greedybear/cronjobs/repositories/sensor.py [hunk 2]
```diff
@@ -31,6 +30,9 @@ def get_or_create_sensor(self, ip: str) -> Sensor | None:
         Returns:
             Sensor object if valid, None if invalid IP format.
         """
+        # TODO: circular import erorr occured
```
Comment (regulartim) [inline greedybear/cronjobs/repositories/sensor.py]: Is there something missing (because there is a TODO comment)?

File: greedybear/admin.py [hunk 3]
```diff
@@ -150,9 +150,21 @@ def general_honeypots(self, ioc):
...
+            name = ioc.autonomous_system.name
+            return f"{asn} ({name})" if name else str(asn)
+        return "-"
+
+    autonomous_system_display.short_description = "Autonomous System"
+
     def get_queryset(self, request):
         """Override to prefetch related sensors and honeypots, avoiding N+1 queries."""
-        return super().get_queryset(request).prefetch_related("sensors", "general_honeypot")
+        return super().get_queryset(request).prefetch_related("sensors", "general_honeypot", "autonomous_system")
```
Comment (regulartim) [inline greedybear/admin.py]: I think for 1:n relations, it is better to use `select_related`. Maybe look that up in the dhjango docs.

File: greedybear/migrations/0041_autonomoussystem_remove_ioc_asn_and_more.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    batch = []
+    for ioc in iocs_qs:
+        ioc.autonomous_system = asn_map[ioc.asn]
+        batch.append(ioc)
+        if len(batch) >= batch_size:
+            IOC.objects.bulk_update(batch, ['autonomous_system'])
+            batch = []
+
+    if batch:
+        IOC.objects.bulk_update(batch, ['autonomous_system'])
```
Comment (regulartim) [inline greedybear/migrations/0041_autonomoussystem_remove_ioc_asn_and_more.py]: I don't think the data migration is fast enough. After the bulk create of all ASes, could we maybe just run something like `IOC.objects.exclude(asn__isnull=True).update(autonomous_system_id=F('asn'))` ? Then we would only need a single query to opdate the IOCs instead of thousands.

File: tests/test_migrations.py (hunk 1 above)
Comment (drona-gyawali) [reply tests/test_migrations.py]: Ah, you’re right  I didn’t notice that . Sorry for this oversight. I’ll move it out so it runs independently and update the PR."

File: greedybear/cronjobs/repositories/sensor.py (hunk 2 above)
Comment (drona-gyawali) [reply greedybear/cronjobs/repositories/sensor.py]: Initially when I was working I was getting circular dependencies error. But now when I moved it to the top of file it is working. not sure but some recent pr solve this

File: greedybear/admin.py (hunk 3 above)
Comment (drona-gyawali) [reply greedybear/admin.py]: Yes you are right, I also verified it from **[doc](https://docs.djangoproject.com/en/6.0/ref/models/querysets/#prefetch-related)**

File: greedybear/migrations/0041_autonomoussystem_remove_ioc_asn_and_more.py (hunk 4 above)
Comment (drona-gyawali) [reply greedybear/migrations/0041_autonomoussystem_remove_ioc_asn_and_more.py]: yes, I thought about that too using a pure DB-level update makes a lot of sense. It would eliminate the need for batching in Python and reduce thousands of queries to just one, which should make the migration much faster. I updated the migration this way .

File: api/views/utils.py [hunk 5]
```diff
@@ -350,6 +350,7 @@ def feeds_response(request=None, iocs=None, feed_params=None, valid_feed_types=N
                     "last_seen": ioc["last_seen"].strftime("%Y-%m-%d"),
                     "feed_type": ioc_feed_type,
                     "destination_port_count": len(ioc.get("destination_ports", [])),
+                    "asn": ioc.get("autonomous_system", ""),
```
Comment (regulartim) [inline api/views/utils.py:353]: After assigning "asn", can you please pop "autonomous_system" from `data_`? Otherwise the asn appears twice in the response.

File: greedybear/cronjobs/repositories/autonomous_system.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        as_obj, created = AutonomousSystem.objects.get_or_create(asn=asn, defaults={"name": name or ""})
+
+        if created:
+            self.log.info(f"Created new ASN {asn} with name '{name or ''}'")
+        elif name and not as_obj.name:
+            as_obj.name = name
+            as_obj.save(update_fields=["name"])
+            self.log.info(f"Updated ASN {asn} name to '{name}'")
+
+        return as_obj
```
Comment (regulartim) [inline greedybear/cronjobs/repositories/autonomous_system.py:40]: What do you think of using a cache for this method, similar to what we are doing in the IOC rrepo with the GerneralHoneyppt instances?

File: greedybear/cronjobs/extraction/utils.py [hunk 7]
```diff
@@ -119,6 +120,7 @@ def iocs_from_hits(hits: list[dict]) -> list[IOC]:
     for hit in hits:
         hits_by_ip[hit["src_ip"]].append(hit)
     iocs = []
+    repo = ASNRepository()  # single instance for this batch
```
Comment (regulartim) [inline greedybear/cronjobs/extraction/utils.py]: Please give "repo" a better name.

File: greedybear/cronjobs/scoring/utils.py [hunk 8]
```diff
@@ -79,7 +79,7 @@ def get_features(iocs: list[dict], reference_day: str) -> pd.DataFrame:
                 "first_seen": ioc["first_seen"],
                 "days_seen": ioc["days_seen"],
                 # CATEGORICAL FEATURES
-                "asn": str(ioc["asn"]),
+                "autonomous_system": str(ioc["autonomous_system"]),
```
Comment (regulartim) [inline greedybear/cronjobs/scoring/utils.py]: I am very skeptical about changes in this function. I would prefer it the key would stay "asn" and the value would be the actual ASN. The ML models don't benefit from the AS name anyway.

File: greedybear/migrations/0043_autonomoussystem_remove_ioc_asn_and_more.py
Comment (regulartim) [inline greedybear/migrations/0043_autonomoussystem_remove_ioc_asn_and_more.py:1]: Nice! :)

**Review by regulartim — CHANGES_REQUESTED** [review]
Hey @drona-gyawali ! I think the cirular import isn't solved...

```
greedybear_qcluster  | 11:55:13 [Q] ERROR Failed 'greedybear.tasks.extract_all' (jig-georgia-yellow-quiet) - cannot import name 'get_ioc_type' from partially initialized module 'greedybear.cronjobs.extraction.utils' (most likely due to a circular import) (/opt/deploy/greedybear/greedybear/cronjobs/extraction/utils.py) : Traceback (most recent call last):
greedybear_qcluster  |   File "/usr/local/lib/python3.13/site-packages/django_q/worker.py", line 103, in worker
greedybear_qcluster  |     res = f(*task["args"], **task["kwargs"])
greedybear_qcluster  |   File "/opt/deploy/greedybear/greedybear/tasks.py", line 11, in extract_all
greedybear_qcluster  |     from greedybear.cronjobs.extract import ExtractionJob
greedybear_qcluster  |   File "/opt/deploy/greedybear/greedybear/cronjobs/extract.py", line 2, in <module>
greedybear_qcluster  |     from greedybear.cronjobs.extraction.pipeline import ExtractionPipeline
greedybear_qcluster  |   File "/opt/deploy/greedybear/greedybear/cronjobs/extraction/pipeline.py", line 4, in <module>
greedybear_qcluster  |     from greedybear.cronjobs.extraction.strategies.factory import ExtractionStrategyFactory
greedybear_qcluster  |   File "/opt/deploy/greedybear/greedybear/cronjobs/extraction/strategies/__init__.py", line 1, in <module>
greedybear_qcluster  |     from greedybear.cronjobs.extraction.strategies.base import *
greedybear_qcluster  |   File "/opt/deploy/greedybear/greedybear/cronjobs/extraction/strategies/base.py", line 4, in <module>
greedybear_qcluster  |     from greedybear.cronjobs.extraction.ioc_processor import IocProcessor
greedybear_qcluster  |   File "/opt/deploy/greedybear/greedybear/cronjobs/extraction/ioc_processor.py", line 4, in <module>
greedybear_qcluster  |     from greedybear.cronjobs.extraction.utils import is_whatsmyip_domain
greedybear_qcluster  |   File "/opt/deploy/greedybear/greedybear/cronjobs/extraction/utils.py", line 11, in <module>
greedybear_qcluster  |     from greedybear.cronjobs.repositories import ASNRepository
greedybear_qcluster  |   File "/opt/deploy/greedybear/greedybear/cronjobs/repositories/__init__.py", line 7, in <module>
greedybear_qcluster  |     from greedybear.cronjobs.repositories.sensor import *
greedybear_qcluster  |   File "/opt/deploy/greedybear/greedybear/cronjobs/repositories/sensor.py", line 4, in <module>
greedybear_qcluster  |     from greedybear.cronjobs.extraction.utils import get_ioc_type
greedybear_qcluster  | ImportError: cannot import name 'get_ioc_type' from partially initialized module 'greedybear.cronjobs.extraction.utils' (most likely due to a circular import) (/opt/deploy/greedybear/greedybear/cronjobs/extraction/utils.py)
```

**Comment by drona-gyawali** [comment]
> Hey @drona-gyawali ! I think the cirular import isn't solved...
> 
> ```
> greedybear_qcluster  | 11:55:13 [Q] ERROR Failed 'greedybear.tasks.extract_all' (jig-georgia-yellow-quiet) - cannot import name 'get_ioc_type' from partially initialized module 'greedybear.cronjobs.extraction.utils' (most likely due to a circular import) (/opt/deploy/greedybear/greedybear/cronjobs/extraction/utils.py) : Traceback (most recent call last):
> greedybear_qcluster  |   File "/usr/local/lib/python3.13/site-packages/django_q/worker.py", line 103, in worker
> greedybear_qcluster  |     res = f(*task["args"], **task["kwargs"])
>
sorry for this,  I thought the error was resolved because the tests were passing. To fix this, we have two options: a simple function-level import b) moving get_ioc_type to a new utility module (helpers.py) to avoid repository dependencies.

Given that this PR already touches over 15+ files, I would  prefer to go with Option 1 for now to keep things clear. I can open a second PR  to handle the refactoring into a utility module if you prefer option b. What you think?

File: greedybear/cronjobs/scoring/utils.py (hunk 8 above)
Comment (drona-gyawali) [reply greedybear/cronjobs/scoring/utils.py]: Thanks for the feedback I have updated the function so the key is now "asn" and the value is the actual ASN, and the old "autonomous_system" key is removed from the response.

File: greedybear/cronjobs/repositories/autonomous_system.py (hunk 6 above)
Comment (drona-gyawali) [reply greedybear/cronjobs/repositories/autonomous_system.py:40]: yes, i added a very basic  in-memory cache for this method, similar in concept to what we do in the IOC repo with GeneralHoneypot. It stores already fetched or created ASNs to avoid duplicate database hits.

**Comment by regulartim** [comment]
> Given that this PR already touches over 15+ files, I would prefer to go with Option 1 for now to keep things clear. I can open a second PR to handle the refactoring into a utility module if you prefer option b. What you think?

Option 1 is fine. We don't need that refactoring right now.

File: greedybear/cronjobs/repositories/autonomous_system.py (hunk 6 above)
Comment (regulartim) [reply greedybear/cronjobs/repositories/autonomous_system.py:40]: Oh, sorry, made a mistake here. I meant the cache at the SensorRepository. Because ASes don't change that often and there is a relatively small amount of them, I think it makes sense to fetch them all, when the repo is instantiated.

File: greedybear/cronjobs/repositories/autonomous_system.py
Comment (regulartim) [inline greedybear/cronjobs/repositories/autonomous_system.py:1]: Maybe we should name the file `as.py` and the class `ASRepository`. Would make more sense, no?

Comment (drona-gyawali) [reply greedybear/cronjobs/repositories/autonomous_system.py:1]: I noticed that (as) is a reserved keyword in Python, so we cannot name the file as.py or use it in imports. I can rename the file to autonomous_system.py and the class to ASRepository instead. Does that sound good to you?

File: greedybear/cronjobs/repositories/autonomous_system.py (hunk 6 above)
Comment (drona-gyawali) [reply greedybear/cronjobs/repositories/autonomous_system.py:40]: thank for feedback I will update that.

File: greedybear/cronjobs/repositories/autonomous_system.py
Comment (regulartim) [reply greedybear/cronjobs/repositories/autonomous_system.py:1]: Ah, right! Yeah, good idea! :)

**Comment by drona-gyawali** [comment]
> Nice, thank you! :) 

Thank you very much for the thoughful review and guidance :)
